// jasmine-allure2-reporter/dist/src/JasmineConsoleReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var JasmineConsoleReporter = /** @class */ (function () {
    function JasmineConsoleReporter() {
    }
    JasmineConsoleReporter.prototype.log = function (message) {
        console.log(new Date().toISOString(), message);
    };
    JasmineConsoleReporter.prototype.suiteStarted = function (suite) {
        this.log("Suite started: " + suite.fullName);
    };
    JasmineConsoleReporter.prototype.suiteDone = function () {
        this.log("Suite ended\n\n");
    };
    JasmineConsoleReporter.prototype.specStarted = function (spec) {
        this.log("Case started: " + spec.description);
    };
    JasmineConsoleReporter.prototype.specDone = function (spec) {
        this.log("Case ended: " + spec.status + "\n\n");
        /*for (const fail of spec.failedExpectations) {
            console.error(fail.message);
            console.error(fail.stack);
        }*/
    };
    return JasmineConsoleReporter;
}());
exports.JasmineConsoleReporter = JasmineConsoleReporter;
