// allure2-js-commons/dist/src/AllureGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TestResultContainer_1 = require("./entities/TestResultContainer");
var Link_1 = require("./entities/Link");
var ExecutableItem_1 = require("./entities/ExecutableItem");
var ExecutableItemWrapper_1 = require("./ExecutableItemWrapper");
var AllureTest_1 = require("./AllureTest");
var AllureGroup = /** @class */ (function () {
    function AllureGroup(runtime, parent) {
        this.runtime = runtime;
        this.parent = parent;
        this.testResultContainer = new TestResultContainer_1.TestResultContainer();
        this.testResultContainer.start = Date.now();
    }
    AllureGroup.prototype.startGroup = function (name) {
        var group = new AllureGroup(this.runtime, this);
        this.testResultContainer.children.push(group.uuid);
        group.name = name || "Unnamed";
        return group;
    };
    AllureGroup.prototype.startTest = function (name) {
        var test = new AllureTest_1.AllureTest(this.runtime, this);
        this.testResultContainer.children.push(test.uuid);
        test.name = name || "Unnamed";
        return test;
    };
    AllureGroup.prototype.endGroup = function () {
        this.testResultContainer.stop = Date.now();
        // TODO: test that children ended
        this.runtime.writeGroup(this.testResultContainer);
    };
    Object.defineProperty(AllureGroup.prototype, "uuid", {
        get: function () {
            return this.testResultContainer.uuid;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AllureGroup.prototype, "name", {
        get: function () {
            return this.testResultContainer.name || "";
        },
        set: function (name) {
            this.testResultContainer.name = name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AllureGroup.prototype, "description", {
        set: function (description) {
            this.testResultContainer.description = description;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AllureGroup.prototype, "descriptionHtml", {
        set: function (descriptionHtml) {
            this.testResultContainer.descriptionHtml = descriptionHtml;
        },
        enumerable: true,
        configurable: true
    });
    AllureGroup.prototype.addLink = function (name, url, type) {
        this.testResultContainer.links.push(new Link_1.Link(name, url, type));
    };
    AllureGroup.prototype.addBefore = function () {
        var result = new ExecutableItem_1.ExecutableItem();
        this.testResultContainer.befores.push(result);
        return new ExecutableItemWrapper_1.ExecutableItemWrapper(result);
    };
    AllureGroup.prototype.addAfter = function () {
        var result = new ExecutableItem_1.ExecutableItem();
        this.testResultContainer.afters.push(result);
        return new ExecutableItemWrapper_1.ExecutableItemWrapper(result);
    };
    return AllureGroup;
}());
exports.AllureGroup = AllureGroup;
