// allure2-js-commons/dist/src/entities/ContentType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ContentType;
(function (ContentType) {
    ContentType["TEXT"] = "text/plain";
    ContentType["XML"] = "application/xml";
    ContentType["CSV"] = "text/csv";
    ContentType["TSV"] = "text/tab-separated-values";
    ContentType["CSS"] = "text/css";
    ContentType["URI"] = "text/uri-list";
    ContentType["SVG"] = "image/svg+xml";
    ContentType["PNG"] = "image/png";
    ContentType["JSON"] = "application/json";
    ContentType["WEBM"] = "video/webm";
    ContentType["JPEG"] = "image/jpeg";
})(ContentType = exports.ContentType || (exports.ContentType = {}));
function typeToExtension(type) {
    switch (type) {
        case ContentType.TEXT: return "txt";
        case ContentType.XML: return "xml";
        case ContentType.CSV: return "csv";
        case ContentType.TSV: return "tsv";
        case ContentType.CSS: return "css";
        case ContentType.URI: return "uri";
        case ContentType.SVG: return "svg";
        case ContentType.PNG: return "png";
        case ContentType.JSON: return "json";
        case ContentType.WEBM: return "webm";
        case ContentType.JPEG: return "jpg";
    }
    throw new Error("Unrecognized extension: " + type);
}
exports.typeToExtension = typeToExtension;
